### ISSUE TITLE
How to retrieve original text from parsed elements?

### ISSUE THREAD

**Issue by Nolence** [petitparser/dart-petitparser#99]
Hi !

Sorry, totally new to parsers. I'm currently just trying to parse `.arb` files so I can translate them with google api. So like turning this

```json
{
  "@@locale": "en",
  "pageHomeInboxCount": "{count, plural, zero{You have no new messages} one{You have 1 new message} other{You have {count} new messages}}",
  "@pageHomeInboxCount": {
    "description": "New messages count on the Home screen",
    "placeholders": {
      "count": {}
    }
  },
}
```
into

```json
{
  "@@locale": "es",
  "pageHomeInboxCount": "{count, plural, zero{No tienes mensajes nuevos} one{Tienes 1 mensaje nuevo} other{Tienes {count} mensajes nuevos}}",
  "@pageHomeInboxCount": {
    "description": "Los mensajes nuevos cuentan en la pantalla de inicio",
    "placeholders": {
      "count": {}
    }
  },
}
```

Copying an [icu_parser](https://gist.github.com/Nolence/ddd1049129d60e2b96199b1d4fc6184c) I found online I correctly parsed the message but now I'd like to goin and change one of the [`List<BaseElement>`](https://gist.github.com/Nolence/7724ade54011b11e4129d15d295aa6fc#file-message_format-dart-L3) (e.g. elements.splice(0, 1, [newElement])) and then go back to the string version with the new addition.

That is, given

```dart
Result<List<BaseElement>> parse(String message) {
  final parsed = (pluralOrGenderOrSelect | simpleText | empty).map(
    (result) {
      final parseResult =
          List<BaseElement>.from(result is List ? result : [result]);
      return parseResult;
    },
  ).parse(message);

  return parsed;
}
```

I'd like to how can I get `message` back?

**Comment by justkawal** [comment]
@Nolence 
I nearly understood what you are trying to do.
But I can help you with one thing is that functions like **splice** and **other useful functions are available in:** https://github.com/justkawal/utility

**I hope this helps.**

**Comment by Nolence** [comment]
Thank you for the help @justkawal ! However if you mean splicing out the strings using pattern matching, I don't think I'd like that idea as I don't think it'd catch every case without some serious thought.

I think my original question can be boiled down to, is there a function that can map the AST back to the original string? I'm thinking if there isn't my preferred second option might be to just override the `toString`'s on the models the parser has and they'll handle it from there. Not sure how robust that would be though:

```dart
class GenderElement extends BaseElement {
  final List<Option> options;

  const GenderElement(String value, this.options)
      : super(ElementType.gender, value);

  @override
  String toString() {
    return '{$value, select, ${options.join(' ')}}';
  }
}
```

Sorry if this code looks out of context. The two links I previously provided have the full code

**Comment by Nolence** [comment]
I actually ended up just getting rid of the intermediary representations and turning things that I intended to translate into `Token`s. The tokens had start and stop and the buffer of the whole string so I just iterated backwards through the tokens and replaced the substring with the translation. See [here](https://github.com/justkawal/arb_translator/pull/2) for full implementation

Thank for the library!
